compiler/z/codegen/OMRMachine.cpp

```diff
@@ -3058,7 +3037,14 @@ OMR::Z::Machine::coerceRegisterAssignment(TR::Instruction
          if (!self()->isAssignable(currentTargetVirtual, currentAssignedRegister))
             {
                {
-               TR_ASSERT(spareReg!=NULL, "coerce reg - blocked, sparereg cannot be NULL.");
+               /**
+                * Register exchange API takes care of exchanging between registers.
+                * In rare case, where virtual register to which real target register
+                * is assigned to is used in memory reference and currentAssignedRegister
+                * is GPR0, we can not do register exchange and in that case we would need
+                * spareReg.
+                */
+               TR_ASSERT_FATAL(spareReg!=NULL, "coerce reg - blocked, sparereg cannot be NULL.");
```

Comment:
I'm a little worried we may end up hitting this. I do think it will be super rare as the register pressure has to be very high and the current assigned register has to be GPR0. I think rather than fatally asserting here, as I cannot think of a way we could handle this scenario, I would say a better solution would be to fail the compilation here instead.

Reply:
Agree with you. I can not really find a way to make sure we reach no such case (Where we have to do register exchange of register used within memory reference with GPR0). Although as you said it is very rare occurrence, it still can occur. In last forced commit, https://github.com/eclipse/omr/compare/8adc1e0485d5e4d8208501833ef7ca2efbe32a82..8d357975be4052b41ac49e9744fa49be218308b7 removed the assert with failing compilation.